sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java

```diff
@@ -304,4 +335,217 @@ private static long parseLong(String key, String value) {
 
     String getFileName();
   }
+
+  /**
+   * Represents a Request parameter, provides information whether is was specified or not (check {@link #isPresent()})
+   * and utility method to nicely handles cases where the parameter is not present.
+   */
+  @Beta
+  public interface Param<T> {
+    boolean isPresent();
+
+    /**
+     * @return the value of the parameter
+     *
+     * @throws IllegalStateException if param is not present.
+     */
+    @CheckForNull
+    T getValue();
+
+    @CheckForNull
+    T or(Supplier<T> defaultValueSupplier);
+  }
+
+  /**
+   * Implementation of {@link Param} where the param is not present.
+   */
+  private enum AbsentParam implements Param<Object> {
+    INSTANCE;
+
+    @SuppressWarnings("unchecked")
+    protected static <T> Param<T> absent() {
+      return (Param<T>) INSTANCE;
+    }
+
+    /**
+     * Always returns true.
+     */
+    @Override
+    public boolean isPresent() {
+      return false;
+    }
+
+    /**
+     * Always throws a {@link IllegalStateException}.
+     */
+    @Override
+    public Object getValue() {
+      throw createGetValueISE();
+    }
+
+    /**
+     * Always returns the value supplied by {@code defaultValueSupplier}.
+     */
+    @Override
+    @CheckForNull
+    public Object or(Supplier<Object> defaultValueSupplier) {
+      return checkDefaultValueSupplier(defaultValueSupplier).get();
+    }
+  }
+
+  /**
+   * Implementation of {@link Param} where the param is present.
+   */
+  private static final class GenericParam<T> implements Param<T> {
+    private final T value;
+
+    private GenericParam(T value) {
+      this.value = value;
+    }
+
+    protected static <T> Param<T> present(T value) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "protected" modifier. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2156)